src/GHC/Util/Brackets.hs

```diff
@@ -0,0 +1,155 @@
+{-# LANGUAGE PackageImports #-}
+{-# LANGUAGE MultiParamTypeClasses , FlexibleInstances, FlexibleContexts #-}
+module GHC.Util.Brackets (Brackets'(..), isApp',isOpApp',isAnyApp',isSection') where
+
+import "ghc-lib-parser" HsSyn
+import "ghc-lib-parser" SrcLoc
+import "ghc-lib-parser" BasicTypes
+
+class Brackets' a where
+  remParen' :: a -> Maybe a -- Remove one paren or nothing if there is no paren.
+  addParen' :: a -> a -- Write out a paren.
+  -- | Is this item lexically requiring no bracketing ever i.e. is
+  -- totally atomic.
+  isAtom' :: a -> Bool
+  -- | Is the child safe free from brackets in the parent
+  -- position. Err on the side of caution, True = don't know.
+  needBracket' :: Int -> a -> a -> Bool
+
+isOpApp', isApp', isSection' :: LHsExpr GhcPs -> Bool
+isApp' (LL _ HsApp{}) = True; isApp' _ = False
+isOpApp' (LL _ OpApp{}) = True; isOpApp' _ = False
+isAnyApp' x = isApp' x || isOpApp' x
+isSection' (LL _ SectionL{}) = True; isSection' (LL _ SectionR{}) = True; isSection' _ = False
+
+instance Brackets' (LHsExpr GhcPs) where
+  -- When GHC parses a section in concrete syntax, it will produce an
+  -- 'HsPar (Section[L|R])'. There is no concrete syntax that will
+  -- result in a "naked" section. Consequently, given an expression,
+  -- when stripping brackets (c.f. 'Hint.Brackets'), don't remove the
+  -- paren's surrounding a section - they are required.
+  remParen' (LL _ (HsPar _ (LL _ SectionL{}))) = Nothing
+  remParen' (LL _ (HsPar _ (LL _ SectionR{}))) = Nothing
+  remParen' (LL _ (HsPar _ x)) = Just x
+  remParen' _ = Nothing
+
+  addParen' e = noLoc $ HsPar noExt e
+
+  isAtom' (LL _ x) = case x of
+      HsVar{} -> True
+      HsUnboundVar{} -> True
+      HsRecFld{} -> True
+      HsOverLabel{} -> True
+      HsIPVar{} -> True
+      -- Note that sections aren't atoms (but parenthesized sections are).
```

Comment:
Are there such things as non-paren sections? Are those not user writable?

Reply:
As far as I can tell, it doesn't seem lexically possible to write a naked section! Ready to be corrected. What I believe to be true is : (1) HSE sees `(+ 1)` as a section, the parens are implicit and the pretty printer is made to write them (2) GHC sees `(+ 1)` as a `HsParen (HsSection ...))`, the parens are explicit and as mentioned above, I don't think it's lexically possible for a user to write a section that isn't parenthesized in this way.

Reply:
TIL - thanks

Reply:
Me too. Brought to you by lots of head scratching here 😉